### PR TITLE
Processing data in Model

### PR DESCRIPTION
**Pull request by Grandi0z**
#### Changes made : 
- Set up associations between models
- Insert data into tables
- Include methods into models

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -1,2 +1,9 @@
 class User < ApplicationRecord
+  has_many :posts
+  has_many :likes
+  has_many :comments
```
Comment (topeogunleye) [inline app/models/user.rb]: Hello @Grandi0z,

Great job on setting up the associations in your User model! However, I noticed a slight discrepancy between the project requirements and your implementation.

As per the ERD Diagram, the User has many posts, comments, and likes.  In your code, you haven't specified the foreign_key for has_many :posts, has_many :comments and has_many :likes.

Remember, you're making good progress! Keep up the good work!

Here's an example of how your code might look like:

```ruby
class User < ApplicationRecord
has_many :posts, foreign_key: 'author_id'
has_many :comments, foreign_key: 'author_id'
has_many :likes, foreign_key: 'author_id'

  def self.three_recent_posts(author_id)
    Post.where(author: author_id).limit(3).order(created_at: :desc)
  end
end
```

The above is an example and may need to be adapted to fit your code.

File: app/models/comment.rb [hunk 2]
```diff
@@ -1,2 +1,8 @@
 class Comment < ApplicationRecord
+  belongs_to :post
+  belongs_to :user
```
Comment (topeogunleye) [inline app/models/comment.rb]: Hello @Grandi0z,

Great job on setting up the associations in your `Comment` model! I noticed that while you've got the association to the `User` model spot on, there's a slight discrepancy with the association to the `Post` model according to the project requirements.

As per the ERD Diagram and the requirements provided, the Comment model should have a specific association with the `Post` model using a custom foreign key.

Remember, you're making great strides! Keep up the excellent work!

Here's how your association for the `Post` model might look:

```ruby
belongs_to :author, class_name: 'User'
belongs_to :post
```

Do ensure the rest of your code aligns with this change.

File: app/models/like.rb [hunk 3]
```diff
@@ -1,2 +1,8 @@
 class Like < ApplicationRecord
+  belongs_to :user
+  belongs_to :post
```
Comment (topeogunleye) [inline app/models/like.rb]: Hello @Grandi0z,

Great job on setting up the associations in your `Like` model! However, I noticed a slight discrepancy between the project requirements and your implementation.

As per the ERD Diagram, the `Like` model should have an association to the `User` model using a custom name and class name. In your code, while you correctly established the `belongs_to :post` association, the association with the `User` model isn't according to the given requirements.

Remember, you're making good progress! Keep up the good work!

Here's how your `Like` model associations might look based on the requirements:

```ruby
class Like < ApplicationRecord
  belongs_to :author, class_name: 'User'
  belongs_to :post
end
```

The above is an example and may need to be adapted to fit the rest of your code structure.

File: app/models/post.rb [hunk 4]
```diff
@@ -1,2 +1,13 @@
 class Post < ApplicationRecord
+  belongs_to :author, class_name: 'User'
+  has_many :comments
+  has_many :posts
```
Comment (topeogunleye) [inline app/models/post.rb]: Hello @Grandi0z,

Great job on setting up the associations in your `Post` model! However, I noticed a slight discrepancy between the project requirements and your implementation.

As per the ERD Diagram, the `Post` model should have associations with comments and likes. In your code, you've correctly set up the `belongs_to :author` and `has_many :comments` associations, but there seems to be a mistake in having `has_many :posts` within the `Post` model itself.

Remember, you're making good progress! Keep up the good work!

Here's an example of how your `Post` model associations might look based on the requirements:

```ruby
class Post < ApplicationRecord
  belongs_to :author, class_name: 'User'
  has_many :comments
  has_many :likes
end
```

The above is an example and may need to be adapted to fit the rest of your code structure.

File: app/models/post.rb [hunk 5]
```diff
@@ -1,2 +1,13 @@
 class Post < ApplicationRecord
+  belongs_to :author, class_name: 'User'
```
Comment (topeogunleye) [inline app/models/post.rb:2]: Hello @Grandi0z,

Your associations in the `Post` model are shaping up well! While reviewing, I had a suggestion that could optimize your code further and make sure the user's post count stays accurate without manual intervention.

👉 **Active Callbacks**: Rails provides a feature called Active Callbacks, which allows methods to be called at different parts of an object's lifecycle. You can utilize the `after_save` callback to automatically update the post count for a user whenever a new post is saved to the database.

Here's a suggested way of doing it:

```ruby
# app/models/post.rb
after_save :update_post_counter

private

def update_post_counter
  author.increment!(:posts_counter)
end
```

By making the `update_post_counter` method private, we ensure it's not callable from outside the class, preserving encapsulation.

This approach will keep the `posts_counter` updated without additional manual steps, making your application more maintainable and less prone to errors in the future.

Remember, this is just a suggestion to enhance your current implementation. You're doing great, and these nuances will further elevate the quality of your code!

Keep it up, and happy coding! 🌟🚀

**Review by topeogunleye — CHANGES_REQUESTED** [review]
Hi @Grandi0z, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

Try-Again

Highlights 🌟

1. Model Organization: You've neatly organized the User, Post, Comment, and Like models inside the app/models/ directory. Good job! 📁

2. Methods: Your recent_post method in the User model, recent_comments in the Post model, as well as counter-updating methods in the Post, Comment, and Like models are well-implemented. 👍

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- For future pull requests, please notice that your branch name `migration` does not follow the convention of being named after a feature or action. For example, you could name your branch `added-migration`. This makes it easier to track the changes and the purpose of each branch.
Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag @topeogunleye in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/post.rb (hunk 5 above)
Comment (Grandi0z) [reply app/models/post.rb:2]: Thank you very much for the tips

File: app/models/user.rb [hunk 6]
```diff
@@ -1,2 +1,14 @@
...
+  has_many :comments, foreign_key: 'author_id'
+
+  def self.three_recent_posts(author_id)
+    Post.where(author: author_id).limit(3).order(created_at: :desc)
+  end
+
+  # def increment_counter_posts
+  #   User.increment_counter(:posts_counter, self.id)
+  #   puts self.id
+  # end
```
Comment (V-Blaze) [inline app/models/user.rb]: - You have done great so far :+1: :+1: Your app meets all the requirements and works as expected. but there are a few things you need to correct before moving to the next milestone. :smile: :wave: 

- [ ] Following the best practices, you are not meant to leave unused code as a comment on your code base. it makes your codebase look unprofessional. I'll suggest you remove these commented-out codes. :smile: :wave:

**Review by V-Blaze — CHANGES_REQUESTED** [review]
## Hi @Grandi0z 

# Almost there 🟡 

1094a3e620ebea4cd6cf2d0aceaa8f52


Great work so far, You have come a long way, keep moving forward👏 👍  
You've done well implementing some changes from the previous reviewers, but there are still some issues I think you need to improve on. 😄 👋 


 ## Highlights
-  Set up association correctly ✔
- correct git flow💯 🥇 ✔
- used Ruby `Active Callbacks` :heavy_check_mark: 

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/post.rb (hunk 5 above)
Comment (topeogunleye) [reply app/models/post.rb:2]: You're very welcome. @Grandi0z

**Review by omar-muhamad — APPROVED** [review]
# Approved ✔
### Hi @Grandi0z,

#### Great job, Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 

### Congratulations! 🎉 

giphy (1)

## (Highlights) Good Points: :+1:
- Your work is well documented. ✔
- The correct workflow is used.  ✔
- All linters checks are ok. ✔

## (Changes Required) Aspects to improve: ♻️

- N/A

## [Optional] suggestions:

- N/A

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

**Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear, and please remember to tag me in your question so I can receive the notification**

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Pull request merged**